mxcubeweb/core/components/harvester.py

```diff
@@ -0,0 +1,494 @@
+        or had and exception (False)
+        """
+        harvester_device = HWR.beamline.harvester
+
+        harvester_device.load_calibrated_pin()
+        harvester_device._wait_sample_transfer_ready(None)
+        print("waiting 40 seconds before mount")
+        # For some reason the Harvester return READY too soon
+        # approximately 40 Second sooner
+        gevent.sleep(40)
+        sample_mount_device = HWR.beamline.sample_changer
+        mount_current_sample = sample_mount_device.load_a_pin_for_calibration()
+
+        if mount_current_sample:
+            try:
+                md = HWR.beamline.diffractometer
+                md._wait_ready()
+
+                sample_drift_x = float(
+                    harvester_device.get_last_sample_drift_offset_x()
+                )
+                sample_drift_y = float(
+                    harvester_device.get_last_sample_drift_offset_y()
+                )
+                sample_drift_z = float(
+                    -harvester_device.get_last_sample_drift_offset_z()
+                )
+
+                motor_pos_dict = {
+                    "kappa": float(
+                        md["HacentringReferencePosition"].get_property("kappa_ref")
+                    ),
+                    "kappa_phi": float(
+                        md["HacentringReferencePosition"].get_property("phi_ref")
+                    ),
+                    "phi": float(
+                        md["HacentringReferencePosition"].get_property("omega_ref")
+                    ),
+                    "phiy": md.phiyMotor.get_value() + sample_drift_x,
+                }
+
+                md.move_motors(motor_pos_dict)
+                md._wait_ready()
+                md.centringFocus.set_value_relative(sample_drift_z, None)
+                md.centringVertical.set_value_relative(sample_drift_y, None)
+
+                md.save_current_motor_position()
+                harvester_device.set_calibrate_state(True)
+
+                logging.getLogger("user_level_log").info(
+                    "Pin Calibration Step 1 Succeed"
+                )
+                logging.getLogger("user_level_log").info(
+                    "User Need to Perform an  3 click centring"
+                )
+                return True
+            except Exception:
+                logging.getLogger("user_level_log").exception("Pin Calibration Failed")
+                return False
+        else:
+            logging.getLogger("user_level_log").error("Pin Calibration Failed")
+            logging.getLogger("user_level_log").error(
+                "Sample Changer could not mount Pin"
+            )
+            return False
+
+    def cancel_calibration(self) -> None:
+        harvester_device = HWR.beamline.harvester
+        harvester_device.set_calibrate_state(False)
+        logging.getLogger("user_level_log").warning("Pin Calibration Canceled")
+
+    def validate_calibration(self) -> bool:
```

Comment:
What do you think about moving this logic to the actual harvester hardware object, it seems mostly to interact with the diffractometer and the harvester and not so much with the UI

Reply:
Maybe I should think of something else,  the harvester hardware object communicate with and only the Harvester . i would prefer not adding MD or other device interaction  to it.

but yes I/we should maybe not put too much on the mxcubeweb core components, and move things down to an Hardware object level.

Reply:
I think we will get alot of clarity from doing that.

Reply:
Yes that would be good, int that case we could make this an Adapter instead of a  Component, that would be much nicer.